Blockzilla/Lib/TrackingProtection/TrackingProtectionPageStats.swift

```diff
@@ -0,0 +1,224 @@
+                    return
+                }
+
+                let json = try Data(contentsOf: URL(fileURLWithPath: path))
+                guard let data = try JSONSerialization.jsonObject(with: json, options: []) as? [[String: AnyObject]] else {
+                    assertionFailure("Blocklists: bad JSON cast.")
+                    return
+                }
+                list = data
+            } catch {
+                assertionFailure("Blocklists: \(error.localizedDescription)")
+                return
+            }
+
+            for rule in list {
+                guard let trigger = rule["trigger"] as? [String: AnyObject],
+                    let filter = trigger["url-filter"] as? String,
+                    let filterRegex = try? NSRegularExpression(pattern: filter, options: []) else {
+                        assertionFailure("Blocklists error: Rule has unexpected format.")
+                        continue
+                }
+
+                guard let loc = filter.range(of: standardPrefix) else {
+                    assert(false, "url-filter code needs updating for new list format")
+                    return
+                }
+                let baseDomain = filter.substring(from: loc.upperBound).replacingOccurrences(of: "\\.", with: ".")
+                assert(!baseDomain.isEmpty)
+
+                // Sanity check for the lists.
+                ["*", "?", "+"].forEach { x in
+                    // This will only happen on debug
+                    assert(!baseDomain.contains(x), "No wildcards allowed in baseDomain")
+                }
+
+                let domainExceptionsRegex = (trigger["unless-domain"] as? [String])?.compactMap { domain in
+                    return wildcardContentBlockerDomainToRegex(domain: domain)
+                }
+
+                // Only "third-party" is supported; other types are not used in our block lists.
+                let loadTypes = trigger["load-type"] as? [String] ?? []
+                let loadType = loadTypes.contains("third-party") ? LoadType.thirdParty : .all
+
+                // Only "font" is supported; other types are not used in our block lists.
+                let resourceTypes = trigger["resource-type"] as? [String] ?? []
+                let resourceType = resourceTypes.contains("font") ? ResourceType.font : .all
+
+                let rule = Rule(regex: filterRegex, loadType: loadType, resourceType: resourceType, domainExceptions: domainExceptionsRegex, list: blockList)
+                blockRules[baseDomain] = (blockRules[baseDomain] ?? []) + [rule]
+            }
+        }
+    }
+
+    func urlIsInList(_ url: URL, whitelistedDomains: [NSRegularExpression]) -> BlocklistName? {
+        let resourceString = url.absoluteString
+        let resourceRange = NSRange(location: 0, length: resourceString.count)
+
+        guard let baseDomain = url.baseDomain, let rules = blockRules[baseDomain] else {
```

Comment:
` let rules = blockRules[baseDomain]` is the grand speedup. If there are no rules for this domain, return immediately, and if there are rules, we get a small list of rules to check. Most domains have <4 associated rules.

Reply:
I like it